src/protocol/sort/compose.rs

```diff
@@ -0,0 +1,127 @@
+/// by K. Chida, K. Hamada, D. Ikarashi, R. Kikuchi, N. Kiribuchi, and B. Pinkas
+/// <https://eprint.iacr.org/2019/695.pdf>
+/// This protocol composes two permutations by applying one secret-shared permutation(sigma) to another secret-shared permutation(rho)
+/// Input: First permutation(sigma) i.e. permutation that sorts all i-1th bits and other permutation(rho) i.e. sort permutation for ith bit
+/// Output: All helpers receive secret shares of permutation which sort inputs until ith bits.
+#[derive(Debug)]
+pub struct Compose<'a, F> {
+    sigma: &'a mut Vec<Replicated<F>>,
+    rho: &'a mut Vec<Replicated<F>>,
+}
+
+impl<'a, F: Field> Compose<'a, F> {
+    #[allow(dead_code)]
+    pub fn new(sigma: &'a mut Vec<Replicated<F>>, rho: &'a mut Vec<Replicated<F>>) -> Self {
+        Self { sigma, rho }
+    }
+    #[embed_doc_image("compose", "images/sort/compose.png")]
+    /// This algorithm composes two permutations (`rho` and `sigma`). Both permutations are secret-shared,
+    /// and none of the helpers should learn it through this protocol.
+    /// Steps
+    /// ![Compose steps][compose]
+    /// 1. Generate random permutations using prss
+    /// 2. First permutation (sigma) is shuffled with random permutations
+    /// 3. Reveal the permutation
+    /// 4. Revealed permutation is applied locally on another permutation shares (rho)
+    /// 5. Unshuffle the permutation with random permutations to negate the effect of shuffling
+    #[allow(dead_code)]
+    pub async fn execute(&mut self, ctx: ProtocolContext<'_, F>) -> Result<(), BoxError> {
+        let mut random_permutations =
+            get_two_of_three_random_permutations(self.rho.len(), &ctx.prss());
+        let mut random_permutations_copy = random_permutations.clone();
+
+        let mut shuffle_sigma = Shuffle::new(self.sigma);
+        shuffle_sigma
+            .execute(ctx.narrow(&ShuffleSigma), &mut random_permutations)
+            .await?;
+
+        let mut perms = reveal_a_permutation(ctx.narrow(&RevealPermutation), self.sigma).await?;
+
+        apply_inv(&mut perms, &mut self.rho);
+
+        let mut unshuffle_rho = Shuffle::new(self.rho);
+        unshuffle_rho
+            .execute_unshuffle(ctx.narrow(&UnshuffleRho), &mut random_permutations_copy)
+            .await?;
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use permutation::Permutation;
+    use rand::seq::SliceRandom;
+    use tokio::try_join;
+
+    use crate::{
+        protocol::{
+            sort::{apply::apply_inv, compose::Compose},
+            QueryId,
+        },
+        test_fixture::{
+            generate_shares, make_contexts, make_world, validate_result_from_shares, TestWorld,
+        },
+    };
+
+    #[tokio::test]
+    pub async fn compose() {
+        const BATCHSIZE: usize = 25;
+        for _ in 0..10 {
+            let mut rng_sigma = rand::thread_rng();
+            let mut rng_rho = rand::thread_rng();
```

Comment:
Why do you need two RNGs?

Reply:
I was using two rngs so that the two obtained permutations are totally independent of each other.